### PR TITLE
OptimizationProblemInterface

### PR DESCRIPTION
**Pull request by gaschler**
Derive optimization problems from a new OptimizationProblemInterface.
Move fix_z into OptimizationProblemOptions.
Move trivial getters to header files.
This will allow mocking the optimization problems.

### CODE REVIEW

File: cartographer/mapping/internal/pose_graph/optimization_problem_interface.h [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+// Implements the SPA loop closure method.
+template <typename NodeDataType, typename SubmapDataType,
+          typename RigidTransformType>
+class OptimizationProblemInterface {
+ public:
+  using Constraint = PoseGraphInterface::Constraint;
+  using LandmarkNode = PoseGraphInterface::LandmarkNode;
+  using NodeData = NodeDataType;
+  using SubmapData = SubmapDataType;
+  using RigidTransform = RigidTransformType;
```
Comment (pifon2a) [inline cartographer/mapping/internal/pose_graph/optimization_problem_interface.h]: Please, remove these typedefs and define them in .cc files only. It won't change a lot here.

Comment (gaschler) [reply cartographer/mapping/internal/pose_graph/optimization_problem_interface.h]: Done

**Comment by gaschler** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user gaschler. Merge queue now has a length of 1.

**Pull request merged**